synapse/telepath.py

```diff
@@ -214,6 +214,70 @@ def __call__(self, *args, **kwargs):
         todo = (self.name, args, kwargs)
         return GenrIter(self.proxy, todo, self.share)
 
+class PipeLine(s_base.Base):
+
+    async def __anit__(self, proxy, genr, name=None):
+
+        await s_base.Base.__anit__(self)
+
+        self.genr = genr
+        self.name = name
+        self.proxy = proxy
+
+        self.count = 0
+        self.genrdone = False
+        self.mustraise = None
+
+        self.link = await proxy.getPoolLink()
+        self.task = self.schedCoro(self._runGenrLoop())
+
+    async def _runGenrLoop(self):
+
+        async for todo in self.genr:
+
+            mesg = ('t2:init', {
+                    'todo': todo,
+                    'name': self.name,
+                    'sess': self.proxy.sess})
+
+            try:
+
+                await self.link.tx(mesg)
+                self.count += 1
+
+            except asyncio.CancelledError as e:
+                return
+
+            except Exception as e:
+                self.mustraise = s_common.err(e)
+                return
+
+        self.genrdone = True
+
+    async def __aiter__(self):
+
+        while not self.isfini:
+
+            if self.mustraise is not None:
```

Comment:
`task` already has the exception, it's done state, and whether it was cancelled. It is literally `self.task.result()` `self.task.done()` and `self.done.cancelled()`

Reply:
(This would also help the currently unhandled case in this loop if the task gets cancelled.)